daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/DelayedTypeSubstitution.scala

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package engine
+package preprocessing
+
+import com.daml.lf.language.Ast.{TVar, Type}
+import data.Ref
+
+// To handle a closed substitution in a lazy way while traversing a term top down.
+private[preprocessing] class DelayedTypeSubstitution private (
+    mapping: Map[Ref.Name, (DelayedTypeSubstitution, Type)]
+) {
+
+  // substitute typ if type is a variable
+  // always return a non variable type
+  @throws[IllegalArgumentException]
+  def apply(typ: Type): (DelayedTypeSubstitution, Type) =
```

Comment:
Does this really solve the issue? You are only normalizing if the variable is the outer most expression. What is the var is nested somewhere inside? Then you have the same issue that you don’t normalize it here and now instead of normalizing it later when you encounter the corresponding variable, you don’t normalize it at all since you no longer recurse here.

I’m sure I’m missing the piece that makes this work, so what is it?

Reply:
The fix I propose consist of 3 point. 

The fix I propose prevent the chaining of variables, which is the only problematic case, complexity wise.
You can easily check that (now) 
- `apply` has a constant complexity 
- `apply` (in ValueTranslator) is applied exactly n times, where n is the size of the value we are translating.
- `apply` (line 34) is applied exactly once for each type variable introduced in the substitution.

Reply:
I’m not quite following, we still recurse after looking up a variable and afaict, we do that once for each occurence of that variable:

1. `subst1.introVars(Seq("a", myComplexDeepType))` to insert `a` into the substitution.
2. Now we recurse. For each occurence of `a` we’ll hit `val (subst1, nonVariableType) = subst0.apply(ty)` and in each of those occurence we’ll recurse on `nonVariableType` meaning we’ll recurse on `myComplexDeepType` as often as we have occurences of `a`.

Is this correct? If so, this seems potentially problematic. What I was thinking of is that in 1, we fully substitute `myComplexDeepType` recursively. In 2, we are then done once we’ve looked up the value in the map, and don’t have to recurse anymore.

Reply:
Oh I think I realized my mistake. We can never avoid the recursion since we have to recurse on the value anyway.

Reply:
We are recursing on both the type **and** the value. 
The function is primitive recursive on the value, so the number time we hit `val (subst1, nonVariableType) = subst0.apply(ty)` is bound by the number of subtrees, in the value tree.

Reply:
I understand what you want to do in 1. I just claim that complexity wise this is useless (and potentially worst).
The current implementation is `O(n+m)` while your proposal will be `O(n+m+l)` where:
- `n` is the size of the value 
- `m` is the number of variable we have to substitute
- `l` is the size of the type.